tsdb/wlog/watcher.go

```diff
@@ -291,14 +291,22 @@ func (w *Watcher) Run() error {
 		return err
 	}
 
+	tail := false
+
 	level.Debug(w.logger).Log("msg", "Tailing WAL", "lastCheckpoint", lastCheckpoint, "checkpointIndex", checkpointIndex, "currentSegment", currentSegment, "lastSegment", lastSegment)
 	for !isClosed(w.quit) {
 		w.currentSegmentMetric.Set(float64(currentSegment))
 		level.Debug(w.logger).Log("msg", "Processing segment", "currentSegment", currentSegment)
 
+		// Reset the value of lastSegment, this is to avoid having to wait too long for between reads if we're reading a
+		// segment that is not the most recent segment after startup. Ignore error here since if there was a problem it would
+		// have been surfaced earlier.
+		_, lastSegment, _ := w.firstAndLast()
```

Comment:
Can we drop the earlier call to `firstAndLast` on line 265 ?  `lastSegment` isn't used until after this point, except in a log message which we can move.

Reply:
I think the earlier call still has some benefit in providing an early exit when there's a problem with the WAL (non-seuqential WAL segments). One option would be to modify the watcher struct to have a pointer to the actual `WL` struct that is created when we open the TSDB, before remote write starts. This would allow us to check `WL.segment` to get the `lastSegment` value we're checking in this function.

Reply:
Is there a lot of work between that early exit and here?
I'm assuming the error check would also be moved.

Reply:
> Is there a lot of work between that early exit and here?

The checkpoint reading, so not 0 work but also not what I'd call "a lot".

I think dropping the earlier call is fine.